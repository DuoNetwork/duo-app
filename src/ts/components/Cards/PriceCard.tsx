import { Tooltip } from 'antd';
import * as d3 from 'd3';
import classAIcon from 'images/ClassA_white.png';
import classBIcon from 'images/ClassB_white.png';
import ethIcon from 'images/ethIcon.png';
import infoIcon from 'images/info.svg';
import moment from 'moment';
import * as React from 'react';
import * as CST from 'ts/common/constants';
import { ColorStyles } from 'ts/common/styles';
import {
	IContractPrice,
	ICustodianContractAddress,
	IDualClassStates,
	ISourceData
} from 'ts/common/types';
import util from 'ts/common/util';
import { SDivFlexCenter } from '../_styled';
import CardTitleSelect from '../Common/CardTitleSelect';
import { SCard, SCardExtraDiv, SCardPriceTag } from './_styled';

interface IProps {
	type: string;
	contractAddress: ICustodianContractAddress;
	locale: string;
	states: IDualClassStates;
	sourceLast: ISourceData<IContractPrice>;
	mobile?: boolean;
}
interface IState {
	source: string;
}
export default class PriceCard extends React.Component<IProps, IState> {
	constructor(props: IProps) {
		super(props);
		this.state = {
			source: ''
		};
	}

	public render() {
		const { locale, states, mobile, contractAddress, type } = this.props;
		const { source } = this.state;
		const last: IContractPrice = CST.API_LIST.includes(source)
			? this.props.sourceLast[source]
			: {
					price: states.lastPrice,
					timestamp: states.lastPriceTime
			};
		const [navA, navB] = CST.API_LIST.includes(source.toUpperCase())
			? util.calculateNav(
					type,
					last.price || 1,
					last.timestamp,
					states.resetPrice,
					states.resetPriceTime,
					states.alpha,
					states.beta,
					states.period,
					states.periodCoupon
			)
			: [states.navA, states.navB];
		const ethChange = last.price / states.resetPrice - 1;
		const tooltipText = (!CST.API_LIST.includes(source.toUpperCase())
			? CST.TT_CTD_NAV
			: CST.TT_EST_NAV)[locale];
		return (
			<SCard
				title={
					<CardTitleSelect
						name={CST.TH_PRICE[locale].toUpperCase()}
						source={''}
						onSelect={(src: string) => this.setState({ source: src })}
					/>
				}
				extra={
					<SCardExtraDiv>
						{last.timestamp
							? CST.TH_LAST_UPDATED[locale] +
							': ' +
							moment(last.timestamp).format('YYYY-MM-DD HH:mm')
							: CST.TH_LOADING[locale]}
					</SCardExtraDiv>
				}
				width={mobile ? '100%' : '590px'}
				margin={mobile ? '20px 0 0 0' : '0 10px 0 0'}
			>
				<SDivFlexCenter horizontal={!mobile} padding="0 10px">
					<SCardPriceTag mobile={mobile} locale={locale}>
						<div className="bg-logo">
							<img src={ethIcon} />
						</div>
						<div className="tag-title">
							<h3>{CST.TH_ETH}</h3>
						</div>
						<div className="tag-content">
							<div style={{ display: 'flex', flexDirection: 'row' }}>
								<div className={'tag-price USD'}>
									{d3.format(',.2f')(last.price)}
								</div>
								<div className={'tag-unit'}>USD</div>
							</div>
							<div className="tag-subtext">
								<div
									style={{
										color:
											ethChange >= 0
												? ColorStyles.TextGreenAlphaL
												: ColorStyles.TextRedAlphaL
									}}
								>
									{d3.format('+.2%')(ethChange)}
								</div>
								<div style={{ marginLeft: 5 }}>
									{CST.TH_SINCE_RESET[locale].toLowerCase()}
								</div>
							</div>
						</div>
					</SCardPriceTag>
					<SCardPriceTag mobile={mobile} locale={locale}>
						<div className="bg-logo">
							<img src={classAIcon} />
						</div>
						<div className="tag-title">
							<a
								className="tag-content"
								href={
									'https://' +
									(__KOVAN__ ? 'kovan.' : '') +
									'etherscan.io/token/' +
									contractAddress.aToken.address
								}
								target="_blank"
							>
								{contractAddress.aToken.code}
							</a>
							<Tooltip title={tooltipText}>
								<img src={infoIcon} />
							</Tooltip>
						</div>
						<div className="tag-content">
							<div style={{ display: 'flex', flexDirection: 'row' }}>
								<div className={'tag-price-1 USD'}>{d3.format(',.6f')(navA)}</div>
								<div className={'tag-unit-1'}>USD</div>
							</div>
							<div className="tag-subtext">
								{type === CST.BEETHOVEN
									? d3.format('.2%')(
											(states.periodCoupon * 365 * 24 * 3600000) /
												states.period || 0
									) +
									' ' +
									CST.TH_PA[locale].toLowerCase()
									: d3.format('.2f')((2 - navA) / (navA || 1)) +
									'x ' +
									CST.TH_LEVERAGE[locale].toLowerCase()}
							</div>
						</div>
					</SCardPriceTag>
					<SCardPriceTag mobile={mobile} locale={locale}>
						<div className="bg-logo">
							<img src={classBIcon} />
						</div>
						<div className="tag-title">
							<a
								className="tag-content"
								href={
									'https://' +
									(__KOVAN__ ? 'kovan.' : '') +
									'etherscan.io/token/' +
									contractAddress.bToken.address
								}
								target="_blank"
							>
								{contractAddress.bToken.code}
							</a>
							<Tooltip title={tooltipText}>
								<img src={infoIcon} />
							</Tooltip>
						</div>
						<div className="tag-content">
							<div style={{ display: 'flex', flexDirection: 'row' }}>
								<div className={'tag-price-2 USD'}>{d3.format(',.6f')(navB)}</div>
								<div className={'tag-unit-2'}>USD</div>
							</div>
							<div className="tag-subtext">
								{d3.format('.2f')((1 + navB) / (navB || 1)) +
									'x ' +
									CST.TH_LEVERAGE[locale].toLowerCase()}
							</div>
						</div>
					</SCardPriceTag>
				</SDivFlexCenter>
			</SCard>
		);
	}
}
