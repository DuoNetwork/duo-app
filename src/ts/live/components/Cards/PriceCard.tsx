import { Tooltip } from 'antd';
import * as d3 from 'd3';
import moment from 'moment';
import * as React from 'react';
import classAIcon from '../../../../images/ClassA_white.png';
import classBIcon from '../../../../images/ClassB_white.png';
import ethIcon from '../../../../images/ethIcon.png';
import infoIcon from '../../../../images/info.svg';
import * as CST from '../../common/constants';
import { ICustodianPrice, ICustodianStates, ISourceData } from '../../common/types';
import util from '../../common/util';
import { SDivFlexCenter } from '../_styled';
import CardTitleSelect from '../Common/CardTitleSelect';
import {
	SCard,
	SCardExtraDiv,
	SCardPriceTag,
} from './_styled';

interface IProps {
	last: ICustodianPrice;
	reset: ICustodianPrice;
	states: ICustodianStates;
	sourceLast: ISourceData<ICustodianPrice>;
}

interface IState {
	source: string;
}

const PriceInfo = (props: {
	icon: string;
	name: string;
	prices: Array<{ value: string; unit: string }>;
	classNamePostfix?: string;
	fromContract: boolean;
}) => {
	const { icon, name, prices } = props;
	const classNamePostfix = props.classNamePostfix || '';
	const tooltipText = props.fromContract
		? 'Nav as currently in Smart Contract'
		: 'Estimated nav based selected ETH price';
	return (
		<SCardPriceTag>
			<div className="bg-logo">
				<img src={icon} />
			</div>

			<div className="tag-title">
				<h3>{name}</h3>
				{name !== 'ETH' ? (
					<Tooltip title={tooltipText}>
						<img src={infoIcon} />
					</Tooltip>
				) : null}
			</div>

			<div className="tag-content">
				<div>
					{prices.map(p => (
						<div key={p.unit} style={{ display: 'flex', flexDirection: 'row' }}>
							<div className={'tag-price' + classNamePostfix + ' ' + p.unit}>
								{p.value}
							</div>
							<div className={'tag-unit' + classNamePostfix}>{p.unit}</div>
						</div>
					))}
				</div>
			</div>
		</SCardPriceTag>
	);
};

export default class PriceCard extends React.Component<IProps, IState> {
	constructor(props: IProps) {
		super(props);
		this.state = {
			source: ''
		};
	}

	public render() {
		const { reset, states } = this.props;
		const { source } = this.state;
		const last: ICustodianPrice = CST.EXCHANGES.includes(source.toUpperCase())
			? this.props.sourceLast[source]
			: this.props.last;
		const [navA, navB] = CST.EXCHANGES.includes(source.toUpperCase())
			? util.calculateNav(
					last.price || 1,
					last.timestamp,
					reset.price,
					reset.timestamp,
					states.alpha,
					states.beta,
					states.period,
					states.periodCoupon
			)
			: [states.navA, states.navB];
		const fromContract = !CST.EXCHANGES.includes(source.toUpperCase());
		return (
			<SCard
				title={
					<CardTitleSelect
						name={CST.TH_PRICE.toUpperCase()}
						onSelect={(src: string) => this.setState({ source: src })}
					/>
				}
				extra={
					<SCardExtraDiv>
						{last.timestamp
							? 'Last Updated: ' + moment(last.timestamp).format('YYYY-MM-DD HH:mm')
							: 'Loading Prices'}
					</SCardExtraDiv>
				}
				width="540px"
				margin="0 10px 0 0"
			>
				<SDivFlexCenter horizontal padding="0 10px">
					<PriceInfo
						icon={ethIcon}
						name={CST.TH_ETH}
						prices={[
							{
								value: d3.formatPrefix(',.2', 1)(last.price),
								unit: 'USD'
							}
						]}
						fromContract={fromContract}
					/>
					<PriceInfo
						icon={classAIcon}
						name={CST.TH_TOKEN_A}
						prices={[
							{
								value: d3.formatPrefix(',.6', 1)(navA),
								unit: 'USD'
							},
							{
								value: d3.formatPrefix(',.8', 1)(navA / (last.price || 1)),
								unit: 'ETH'
							}
						]}
						fromContract={fromContract}
						classNamePostfix="-1"
					/>
					<PriceInfo
						icon={classBIcon}
						name={CST.TH_TOKEN_B}
						prices={[
							{
								value: d3.formatPrefix(',.6', 1)(navB),
								unit: 'USD'
							},
							{
								value: d3.formatPrefix(',.8', 1)(navB / (last.price || 1)),
								unit: 'ETH'
							}
						]}
						fromContract={fromContract}
						classNamePostfix="-2"
					/>
				</SDivFlexCenter>
			</SCard>
		);
	}
}
