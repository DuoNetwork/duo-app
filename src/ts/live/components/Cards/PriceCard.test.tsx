import { shallow } from 'enzyme';
import * as React from 'react';
import PriceCard from './PriceCard';

describe('AuthCard Test', () => {
	describe('User Login', () => {
		const last = {
			address: '0x00476E55e02673B0E4D2B474071014D5a366Ed4E',
			price: 203.59500509278894,
			timestamp: 1539846000000
		};
		const locale = 'EN';
		const reset = {
			address: '0x00476E55e02673B0E4D2B474071014D5a366Ed4E',
			price: 252.93877621546696,
			timestamp: 1534258801000
		};
		const sourceLast = {
			gdax: { address: '0x0', price: 203.59, timestamp: 1539844701135 },
			gemini: { address: '0x0', price: 203.51, timestamp: 1539844427990 },
			kraken: { address: '0x0', price: 203.76, timestamp: 1539844698485 }
		};
		const states = {
			addrPoolLength: 6,
			adminCoolDown: 3600,
			alpha: 1,
			beta: 1,
			createCommRate: 0.01,
			duoBalance: 2482128.9762267205,
			ethBalance: 742.408092589115,
			ethDuoFeeRatio: 800,
			feeAccumulated: 87.06788278541666,
			iterationGasThreshold: 65000,
			lastAdminTime: 0,
			limitLower: 0.25,
			limitPeriodic: 1.035,
			limitUpper: 2,
			navA: 1.026367,
			navB: 0.5834692468501522,
			nextResetAddrIndex: 0,
			numOfPrices: 0,
			period: 3600,
			periodCoupon: 0.000017,
			preResetWaitingBlocks: 10,
			priceFeedTimeTol: 60,
			priceFeedTol: 0.01,
			priceTol: 0.05,
			priceUpdateCoolDown: 3000,
			redeemCommRate: 0.01,
			state: 'Trading',
			totalSupplyA: 82880.4753362674,
			totalSupplyB: 82880.4753362674,
			usersLength: 295
		};
		it('Test Snapshot', () => {
			const wrapper = shallow(
				<PriceCard
					last={last}
					locale={locale}
					reset={reset}
					sourceLast={sourceLast}
					states={states}
				/>
			);
			expect(wrapper).toMatchSnapshot();
		});
	});
});
