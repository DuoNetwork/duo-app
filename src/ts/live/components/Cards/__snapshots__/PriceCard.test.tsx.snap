// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthCard Test User Login Test Snapshot 1`] = `
<Styled(Card)
  extra={
    <styled.div>
      Last Updated: 2018-10-18 15:00
    </styled.div>
  }
  margin="0 10px 0 0"
  title={
    <CardTitleSelect
      name="PRICE"
      onSelect={[Function]}
      source=""
    />
  }
  width="540px"
>
  <styled.div
    horizontal={true}
    padding="0 10px"
  >
    <styled.div
      locale="EN"
    >
      <div
        className="bg-logo"
      >
        <img
          src={null}
        />
      </div>
      <div
        className="tag-title"
      >
        <h3>
          ETH
        </h3>
      </div>
      <div
        className="tag-content"
      >
        <div
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <div
            className="tag-price USD"
          >
            203.60
          </div>
          <div
            className="tag-unit"
          >
            USD
          </div>
        </div>
        <div
          className="tag-subtext"
        >
          <div
            style={
              Object {
                "color": "rgba(255,94,94,.6)",
              }
            }
          >
            -19.51%
          </div>
          <div
            style={
              Object {
                "marginLeft": 5,
              }
            }
          >
            since reset
          </div>
        </div>
      </div>
    </styled.div>
    <styled.div
      locale="EN"
    >
      <div
        className="bg-logo"
      >
        <img
          src={null}
        />
      </div>
      <div
        className="tag-title"
      >
        <a
          className="tag-content"
          href="https://kovan.etherscan.io/tokens?q=0x8e9962286823F21960D849CCC52F8c4a09a4b30f"
          target="_blank"
        >
          Token A
        </a>
        <Tooltip
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          mouseEnterDelay={0.1}
          mouseLeaveDelay={0.1}
          placement="top"
          prefixCls="ant-tooltip"
          title="Nav as currently in Smart Contract"
          transitionName="zoom-big-fast"
        >
          <img
            src={null}
          />
        </Tooltip>
      </div>
      <div
        className="tag-content"
      >
        <div
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <div
            className="tag-price-1 USD"
          >
            1.026367
          </div>
          <div
            className="tag-unit-1"
          >
            USD
          </div>
        </div>
        <div
          className="tag-subtext"
        >
          14.89% p.a.
        </div>
      </div>
    </styled.div>
    <styled.div
      locale="EN"
    >
      <div
        className="bg-logo"
      >
        <img
          src={null}
        />
      </div>
      <div
        className="tag-title"
      >
        <a
          className="tag-content"
          href="https://kovan.etherscan.io/tokens?q=0x1575e11F5DA9067A577175f898A92e9B4BfbE060"
          target="_blank"
        >
          Token B
        </a>
        <Tooltip
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          mouseEnterDelay={0.1}
          mouseLeaveDelay={0.1}
          placement="top"
          prefixCls="ant-tooltip"
          title="Nav as currently in Smart Contract"
          transitionName="zoom-big-fast"
        >
          <img
            src={null}
          />
        </Tooltip>
      </div>
      <div
        className="tag-content"
      >
        <div
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
            }
          }
        >
          <div
            className="tag-price-2 USD"
          >
            0.583469
          </div>
          <div
            className="tag-unit-2"
          >
            USD
          </div>
        </div>
        <div
          className="tag-subtext"
        >
          2.76x leverage
        </div>
      </div>
    </styled.div>
  </styled.div>
</Styled(Card)>
`;
